actions/contacts/upsert.go

```diff
@@ -0,0 +1,63 @@
+package contacts
+
+import (
+	"encoding/json"
+	"errors"
+	"net/http"
+
+	"github.com/bitcoin-sv/spv-wallet/engine"
+	"github.com/bitcoin-sv/spv-wallet/mappings"
+	"github.com/bitcoin-sv/spv-wallet/models"
+	"github.com/bitcoin-sv/spv-wallet/server/auth"
+	"github.com/gin-gonic/gin"
+)
+
+// upsert will add new or modified existing contact
+// Upsert contact godoc
+// @Summary		Upsert contact
+// @Description	Add or update contact. For new contact send request to add current user as contact
+// @Tags		Contact
+// @Produce		json
+// @Param		UpsertContact body contacts.UpsertContact true "Full name and metadata needed to add/modify contact"
+// @Success		201
+// @Router		/v1/contact/{paymail} [PUT]
+// @Security	x-auth-xpub
+func (a *Action) upsert(c *gin.Context) {
+	requesterPubKey := c.GetString(auth.ParamXPubKey)
+	cPaymail := c.GetString("paymail")
+
+	var req UpsertContact
+	if err := json.NewDecoder(c.Request.Body).Decode(&req); err != nil {
+		c.JSON(http.StatusBadRequest, err.Error())
+		return
+	}
+
+	if err := req.validate(); err != nil {
+		c.JSON(http.StatusBadRequest, err.Error())
+		return
+	}
+
+	contact, err := a.Services.SpvWalletEngine.UpsertContact(
+		c.Request.Context(),
+		req.FullName, cPaymail,
+		requesterPubKey,
+		engine.WithMetadatas(req.Metadata))
+
+	if err != nil && !errors.Is(err, engine.ErrAddingContactRequest) {
+		c.JSON(http.StatusUnprocessableEntity, err.Error())
+		return
+	}
+
+	contract := &models.CreateContactResponse{
+		Contact: mappings.MapToContactContract(contact),
+	}
+
+	if err != nil {
+		ai := make(map[string]string)
+		ai["warning"] = err.Error()
+
+		contract.AdditionalInfo = ai
```

Comment:
Suggestion - We could add a method to the contact, like 'AddAdditionalInfo(key, value string)' which would create a map if necessary and add an element.

Reply:
Done